rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,345 @@
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true
+  },
+  "affected_records": 3
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  },
+  "affected_records": 0
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
+will be under the `options.record_session.db` role option. It will support the
+following values:
+- `on`: Enables session recording.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
+
+The value will be set to `on` by default, keeping the current behavior.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      db: on|off
+```
+
+### Player
+
+To play the session recording, we will rely on the existing player used by SSH
+sessions. This means we don't need to introduce any new player on `tsh` or the
+Web UI. We're going to convert database recording events into `SessionPrint`
+events, which can be rendered by the players.
+
+This conversion will be agnostic to the database protocol. This will make it
+easier to keep the session recording play consistent across protocols. However,
+protocol-specific messages will still require special handling, otherwise the
+player won't be able to present them.
+
+It will present the session recording events in different text formats:
+
+- Status (result): Show the user message. If none is present, show "SUCCESS" in
+  case of success or "ERROR" in case of error.
+- The player will define if there is need to display the number of affected
+  records. Protocol-specific variations can be added for this. For example, if
+  the query was a `SELECT` the player might display it as "Returned X rows"
+  after the status.
+
+Example of player visualization:
+
+```code
+mydatabse=# SELECT id, name FROM events;
+SUCCESS
+(Returned 3 rows)
+
+mydatabase=# INSERT INTO events (name) VALUES ('session.query');
+SUCCESS
+(1 row inserted)
+
+mydatabase=# SELECT with_error;
+ERROR: column "with_error" does not exist
+
+mydatabase=# ALTER SYSTEM SET client_min_messages = 'notice';
+SUCCESS
+
+(session end)
+```
+
+#### Legacy session recordings
+
+Older sessions records will not contain the results, however, the player will
+still be able to play them. For this, the player will only present the queries
+executed, following the same format described earlier.
+
+Example:
+
+```code
+mydatabse=# SELECT id, name FROM events;
+
+mydatabase=# INSERT INTO events (name) VALUES ('session.query');
+
+mydatabase=# SELECT with_error;
+
+mydatabase=# ALTER SYSTEM SET client_min_messages = 'notice';
+
+(session end)
+```
+
+### Performance considerations
+
+Recording queries and responses will increase storage requirements and
```

Comment:
All audit events are already being recorded in the database session recordings, no? We're just adding ability to replay them.

Reply:
> All audit events are already being recorded in the database session recordings, no?

Yes, but we're adding the response events, so we can expect the session recordings to increase. There will be two instead of a single event per query/command.